Step3 문자열 계산기

AnKyungMoo opened the pull request:
리뷰 요청합니다~!

AnKyungMoo committed:
feat: 코드리뷰 반영

AnKyungMoo committed:
docs: step3 기능 정의

AnKyungMoo committed:
feat: 덧셈 기능 추가

AnKyungMoo committed:
feat: 사칙연산 구현

AnKyungMoo committed:
test: 덧셈 테스트 추가

AnKyungMoo committed:
test: 사칙연산 테스트 추가

- 뺄셈, 곱셉, 나눗셈 테스트 추가
- 나눗셈은 0으로 나눴을 경우 exception 나오는지 확인

AnKyungMoo committed:
feat: 입력 기능 추가

AnKyungMoo committed:
test: input 테스트 추가

- 빈문자열이나 null일 때 illegalArgumentException이 나옴

AnKyungMoo committed:
feat: 입력에 사칙연산 기호 검증 추가

AnKyungMoo committed:
feat: 사칙연산을 모두 포함하는 기능 추가

- 각 연산을 private으로 변경
- 테스트코드에 연산 결과 확인하는 방법 변경
- input 할 때 마다 expression을 비워주도록 변경

AnKyungMoo committed:
refactor: 연산자를 enum으로 변경

AnKyungMoo committed:
style: 미사용 괄호 제거

namjackson commented on app/src/test/java/edu/nextstep/camp/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,203 @@
+package edu.nextstep.camp.calculator
```
Calculator domain관련 테스트도 도메인 모듈에서 작성해주세요!

AnKyungMoo replied:
<img width="593" alt="image" src="https://user-images.githubusercontent.com/22673963/179349636-984d996a-7902-4467-983e-b534dcde5814.png">
이렇게 나오는데 혹시 어떻게 파일을 만들어야할까요ㅜ..
키워드가 감이 잡히지 않아서 검색해도 잘 나오지가 않네요..ㅠㅠ

AnKyungMoo replied:
해결했습니다!ㅎㅎ;;

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Operation.kt:
```diff
@@ -0,0 +1,22 @@
+                "+" -> PLUS
+                "-" -> MINUS
+                "*" -> MULTIPLY
+                "/" -> DIVIDE
```
연산자들은 이미 enum클래스에 정의 되어 있기 때문에, 따로 작성하지 않으셔도 될거같아요! 
또, 이러한 구조의 경우, enum에서 연산자가 추가될때마다, 관리가 필요한구조로 보여지네요! 
`values`로 find 등의 함수 활용해보시는것도 좋을거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+package edu.nextstep.calculator.domain
+
+class Calculator {
```
해당 클래스에 문자열, 연산자 등의 많은 책임이 들어있는것 같아요! 
책임을 나눠서, 서로 협력할수 있도록 변경해보는건 어떨까요?

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Operation.kt:
```diff
@@ -0,0 +1,22 @@
+package edu.nextstep.calculator.domain
+
+enum class Operation(value: String) {
```
연산의 의미보다는 연산자라는 의미의 Operator라는 이름이 명확할거 같아요! 

추가로 Operation클래스 에 대한 테스트 코드가 없네요! 
함께 작성해주셔도 좋을거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+    private fun plus(first: Int, second: Int): Int = first + second
+    private fun minus(first: Int, second: Int): Int = first - second
+    private fun multiply(first: Int, second: Int): Int = first * second
+    private fun divide(first: Int, second: Int): Int = first / second
```
calculateExpression, 연산 함수의 역할은 
연산자 클래스에 조금더 어울리는거 같아요! 
분리해보시는 것을 추천합니다!

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+
+    private fun isValidExpression(expressionContents: List<String>) {
+        expressionContents.forEach { content ->
+            if (isNumber(content)) {
```
for문안에 if문이 중첩되어 있네요! 
최대한 함수로 분리해보시는 것을 추천드립니다!
 
그리고 해당 함수로는 유효성 검사가 부족해보여요!

namjackson commented on app/src/test/java/edu/nextstep/camp/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    @Test
+    fun `빈 문자열이나 null을 input하면 IllegalArgumentException이 나온다`() {
```
@Test(expected = IllegalArgumentException::class)
어노테이션으로도 assertThrows를 대체할 수 있어요! 👍

AnKyungMoo replied:
어노테이션으로 Exception을 체크했을 때 장점이 어떤것인가요?
어떨 때 나눠서 쓰는게 좋을지 몰라서 질문드립니다!

namjackson replied:
기능상 동일하긴 하고, 현재는 테스트코드 구현부가 간단하여 큰 의미는 없을 수 있는데,

개인적으로 테스트 코드가 많아지고, 테스트 구현부가 길어지다보면, 
구현부를 지속적으로 확인하기가 힘든 부분이 있을수 있어요
테스트 케이스명만 보게되는데, 어노테이션으로 해주면 가독성이 좋아지는 효과가 있는거 같아요!

AnKyungMoo replied:
답변 감사합니다ㅎㅎ

namjackson commented on app/src/test/java/edu/nextstep/camp/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,203 @@
+import org.junit.Before
+import org.junit.Test
+
+class CalculatorTest {
```
테스트 코드를 잘 작성해주셨는데, 

테스트 범위에 에 예측하지 못하는 케이스가 존재하네요! 
- `"+  + 1" ` 피연산자 자리에 연산자가 들어갈 케이스
- `"1+1" ` 띄어쓰기 안된 케이스
보통 테스트 코드의 경우, 정상 동작을 테스트하는 happy case도 중요하지만, 
고의로 실패시키는 unHappy case 또한 중요한거 같아요!  

이런 부분은 근본적으로 Calculator에 많은 책임이 있어서 
CalculatorTest에서 많은 테스트 케이스가 존재하기 때문에, 
미쳐 신경쓰지못하거나 놓치는 부분이라고 생각이 됩니다!

Calculator의 책임을 나눠보고, 테스트 케이스를 보완해주시는것도 좋을거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+        get() = _expressionContents
+    private val _expressionContents = mutableListOf<String>()
+
+    fun input(expression: String?) {
```
이번 미션의 요구사항 중에는
Calculator에서 expressionContents을 가지고 있을 필요는 없는거 같아요! 
Calculator 객체가 동시에 사용되는 케이스등에 문제가 생길수 있기때문에,
불필요한 프로퍼티를 보유하는건 지양하는게 좋아보여요! 

input의 경우에도 모호한 함수라고 생각이 들어요.
계산기의 어떤부분을 input 해주는건지, 완성된 표현식만 넣어야할지, 
표현식중 일부만 넣어야할지에 대한 부분도 구현부를 보지않으면 모를거같아요! 
또, calculate를 되지 않지만, validation은 해주기 때문에, 해당 함수를 사용하는 사람이 모호하게 생각할거 같아요! 

input을 사용하는거보다 calculate에 파라미터로 전달해주는 것이 좋아보여요!

namjackson reviewed (changes requested):
안녕하세요 경무님!  
Calculator에 관련해서 몇가지 코멘트 남겼습니다.
반영 이후에 다시 리뷰 요청을 해주세요!

bran-ahn committed:
feat: 코드리뷰 적용

- calculator와 operator의 책임 분리
- 변수명 변경

bran-ahn committed:
feat: 코드리뷰 반영

- 테스트 코드 수정
- Operator 정상적으로 가져오도록 수정

bran-ahn committed:
test: 테스트 케이스 추가

bran-ahn committed:
feat: domain 모듈에 테스트 dependency 추가

bran-ahn committed:
test: 테스트 코드 역할 분리

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Operator.kt:
```diff
@@ -0,0 +1,29 @@
+    companion object {
+        fun fromValue(value: String) = values().associateBy(Operator::value)[value] ?: UNDEFINED
+
+        fun calculateExpression(first: Int, second: Int, operation: Operator): Int {
```
Operator 파라미터를 넘겨주는 Operator의 companion object 함수를 선언하는 것은 불필요해보여요,
해당 함수를 Operator 클래스의 로컬 함수로 선언해주세요!

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Operator.kt:
```diff
@@ -0,0 +1,29 @@
+        private fun plus(first: Int, second: Int): Int = first + second
+        private fun minus(first: Int, second: Int): Int = first - second
+        private fun multiply(first: Int, second: Int): Int = first * second
+        private fun divide(first: Int, second: Int): Int = first / second
```
enum클래스를 활용하면 
위의 사칙연산 함수들을 더 간단하게 사용할 수 있을거 같아요!
https://kotlinlang.org/docs/enum-classes.html

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,62 @@
+package edu.nextstep.calculator.domain
+
+class Calculator {
```
문자열 검증 기능이  완벽하지 않은거 같아요
현재의 로직이라면, "1 + 2 +"와 같은 케이스를 잡아주지 못하고 있어요 🥲 
숫자+(연산자+숫자) 와 같은 정규표현식 활용해 보셔도 좋을거 같아요

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,62 @@
+        return expressionContents
+    }
+
+    private fun isValidExpression(expressionContents: List<String>) {
```
바로 위에서 언급했지만, 문자열 과련된 함수는 테스트가 필요한거 같아요! 
테스트가 필요한 함수에 private이 되어있어 테스트가 힘들다면,

문자열 가공 / 검증 관련된 클래스의 분리를 통해 테스트를 작성해보세요! 
Calculator에서의 문자열 관련된 책임을 덜어줄수 있습니다!

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/Operator.kt:
```diff
@@ -0,0 +1,29 @@
+    UNDEFINED("UNDEFINED");
+
+    companion object {
+        fun fromValue(value: String) = values().associateBy(Operator::value)[value] ?: UNDEFINED
```
UNDEFINED을 활용하는 방법도 좋지만, 
해당 값이 없을 경우, null로 리턴하는 방법도 있어요!

namjackson reviewed (changes requested):
피드백 적용을 잘해주셨습니다.
몇 가지 코멘트을 다시 남겼습니다. 🙏 
피드백 적용 후, 다시 리뷰요청해주세요!

bran-ahn committed:
feat: 코드리뷰 반영

- 수식 검증 관련 로직을 다른 클래스로 분리
- 수식 검증 정규표현식 개선
- Calculator와 ExpressionValidation은 상태를 가지지 않으므로 object로 변경
- Operator의 함수 개선

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/ExpressionValidation.kt:
```diff
@@ -0,0 +1,11 @@
+package edu.nextstep.calculator.domain
+
+object ExpressionValidation {
+    private val expressionRegex = Regex("-?\\d+( [+\\-*/] -?\\d+)*")
```
정규표현식 활용 👍

namjackson commented on domain/src/test/kotlin/edu/nextstep/calculator/domain/ExpressionValidationTest.kt:
```diff
@@ -0,0 +1,36 @@
+        ExpressionValidation.isValidExpression("1")
+        ExpressionValidation.isValidExpression("0")
+        ExpressionValidation.isValidExpression("-1")
+        ExpressionValidation.isValidExpression("100002")
```
미션진행하시면서, 여유가 있으시면,
여러케이스를 동시에 진행해야할때,
ParameterizedTest를 적용해보셔도 좋을거같아요!

namjackson commented on domain/src/main/java/edu/nextstep/calculator/domain/ExpressionValidation.kt:
```diff
@@ -0,0 +1,11 @@
+package edu.nextstep.calculator.domain
+
+object ExpressionValidation {
```
Validation이라는 이름보단, Validator라는 이름도 괜찮은거같아요!

namjackson reviewed (approved):
피드백부분 모두 적용을 잘해주셨네요! 
3단계 미션 고생하셨습니다! 
다음 미션을 진행 해주세요!
